avocado/core/test.py

```diff
@@ -337,13 +429,74 @@ def datadir(self):
         contained in a given test file.
         """
+        # TODO: on removal of Test.datadir, move its code over to
+        # get_file_data_dir
         # Maximal allowed file name length is 255
         if (self.filename is not None and
                 len(os.path.basename(self.filename)) < 251):
             return self.filename + '.data'
         else:
             return None
 
+    def get_data(self, filename):
+        """
+        Returns the path of a data file, if one exists
+
+        This method respects the existence of directories in the
+        following order:
+
+        1) "$(TEST_FILENAME).data/$(TEST)/$(VARIANT-$HASH)", if a
+           variant exists
+        2) "$(TEST_FILENAME).data/$(TEST)/$(VARIANT)", if a
+           variant exists
+        3) "$(TEST_FILENAME).data/$(TEST)", if no variant is used
+        4) "$(TEST_FILENAME).data", as a fallback for files common
+           to all tests hosted on a single file
+
+        :param filename: the relative name of the file to be searched
+                         on the the possible data directories
+        :type filename: str
+        :returns: either the path to the data file or None
+        :rtype: str or None
+        """
+        log_fmt = 'DATA (filename=%s) => %s (%s)'
+        datadir = get_file_data_dir(self)
+        if datadir is None or not os.path.isdir(datadir):
+            self.log.debug(log_fmt, filename, None,
+                           "no data dir exists")
+            return None
+
+        variant_data_dir = get_variant_data_dir(self)
+        if variant_data_dir is not None:
+            variant_filename = os.path.join(variant_data_dir, filename)
+            if os.path.exists(variant_filename):
+                self.log.debug(log_fmt, filename, variant_filename,
+                               "found at variant-level data dir")
+                return variant_filename
+
+        variant_data_dir = get_simplified_variant_data_dir(self)
+        if variant_data_dir is not None:
+            variant_filename = os.path.join(variant_data_dir, filename)
+            if os.path.exists(variant_filename):
+                self.log.debug(log_fmt, filename, variant_filename,
+                               "found at simplified (name-only, no hash) "
+                               "variant-level data dir")
+                return variant_filename
+
+        test_data_dir = get_test_data_dir(self)
+        test_filename = os.path.join(test_data_dir, filename)
+        if os.path.exists(test_filename):
+            self.log.debug(log_fmt, filename, test_filename,
+                           "found at test-level data dir")
+            return test_filename
+
+        file_filename = os.path.join(datadir, filename)
+        if os.path.exists(file_filename):
+            self.log.debug(log_fmt, filename, file_filename,
+                           "found at file-level data dir")
+            return file_filename
+        return None
```

Comment:
Imagine there is a `datadir`, but no file is matching. To get here you call the:
* get_file_data_dir - 4 times
* get_variant_data_dir - 3 times
* ...

I think it'd be better to implement an iterator, instead of standalone methods to get next possible data-dir. The only difference would be it'd store the already discovered dirs and it'd yield the next one until the end. The `get_data` would then simply:

```
for datadir in self._data_dirs:
    path = os.path.join(datadir, filename)
    if os.path.exists(path):
        return path
```

and the `Test._data_dirs` would be lazily-evaluated generator. As a benefit people would be able to override it (if needed) and extend the data-dirs concept of their custom locations.

Reply:
Also perhaps it makes sense to allow calling `get_data(None)` which should report the first none-None location in the hierarchy. That would be used by stdout/err recorder, or even by advanced users for custom datadir handling.

Reply:
Implement first, optimize later.  If the idea is sound, we can work on speedier implementations later.  So, getting back to the ideas part, I think the second comment here is the relevant one.  

IIUC, you propose the use of `get_data()` for another purpose, which is to retrieve the *directory* and not a data file, right?  I don't think we should be "`ioctl()`ing" functions/methods, making them suit different purposes.  If test *writers* need access to the directory, then we can think about an API for that.

Also, the current idea is that empty data-dirs mean nothing to the test.  These datadir should be writable by the test, so it doesn't make any sense to me to return the first data dir in existence in the hierarchy.

Reply:
I knew I shouldn't have start with the optimization, but that was not the main topic. My point is I'd really prefer unnamed hierarchy, let's say list/iterable of related datadirs to 3 explicitly named dirs. Those 3 definitions work for INSTRUMENTED tests only, but there are (and suppose to be) more test types and hierarchy is all people should need from the concept of the test.

Even later we thought about optimizing the storage (putting shared files upper in the hierarchy) we should not need named purposes. Simply we'd ascend the hierarchy and look for the most common denominator (or how is it called in English).

So the main point (of this whole review) is whether you'd consider having the hierarchy defined by `get_data`, which should operate on top of the list of dirs starting from `self.datadir` and add the variants that are available (unless they are None).

As for the proposed `get_data(None)` I agree it's probably not needed to share the same function, anyway I believe we need some API to get all related dirs. It'd be definitely used by stdout/err recorder and I bet users would find other usages as well.

Overall I'd go with something like `self.get_datadirs` which should return list/iterable with all related dirs. That wold be THE function used by users to extend this concept, should they need to. Then there should be the `self.get_data`, which should use the `self.get_datadirs` and should look for the files available on different levels.

Personally I don't need the names for the datadirs, because the names only make sense for that particular test type. Other tests might use different hierarchy, for example let's say we want to improve Avocado-vt to use more Avocado code, Avocado-vt uses `deps` which can come from test-providers or from Avocado (or several other places). Also we can expand it of `.data` dirs which would be associated with the test type (therefor the filename of the executed tests). In such example the `datadir` would be the same, only related to the discovered file instead of the `avocado_vt/test.py` and the rest of the hierarchy would be related to installed providers and avocado-vt shared data and not to the method/class/variant names...

Reply:
Actually that would be a nice Avocado-vt addition and it'd allow the stdout/err collection. Only the hierarchy should be:

* test_type_filename/name_from_subtests.cfg/name
* test_type_filename/name_from_subtests.cfg
* test_type_filename
* test_provider/deps
* avocado_vt_shared/deps

What do you think? Does this fit your proposed hierarchy? In what I described I'd just modify the `self.get_datadirs` and the `get_data` would automatically used that...